### ISSUE TITLE
Training strategies

### ISSUE THREAD

**Issue by DhairyaLGandhi** [DhairyaLGandhi/ResNetImageNet.jl#21]
Given that we now have a vanilla DDP implementation which can scale on a single node, it might be worth thinking about the various scheduling strategies we can use to speed up training. One of them being gradient overlapping or bucketing. Here, we intentionally do not update the model but only pull out gradients over several batches and update the model every N iterations. 

The advantages are:
* This way we can reduce the reliance on synchronisation often
* and make it so the gradient sync can also be handled asynchronously to the gradient collection tasks

This is similar to how Pytorch also implements some of the PyTorch distributed code (see section 3.2.4 in [their paper](https://www.vldb.org/pvldb/vol13/p3005-li.pdf)).

We can of course go further to hide latency where possible. Currently, resnet 34 trains at about 650 fps on 4 V100s, reducing the copies by not broadcasting the gradients gets an additional bump. Most importantly, there seem to be places where the CPU pauses GPU execution which may be linked to type stability.

**Comment by DhairyaLGandhi** [comment]
This should also be thinking of a good case for scheduling. Maybe FluxML/Optimisers.jl#15
